Stop using g_unix_signal_add() to avoid threads

allisonkarlitskaya opened the pull request:
g_unix_signal_add() is implemented via an implicitly-created worker
thread in GLib.  This thread creates an eventfd, and conflicts with
conmon's general approach to taking full ownership of all open fds in
the process.

Stop using g_unix_signal_add() and switch to using g_unix_fd_add() on a
signalfd() as an alternative approach.

Fixes #330


I tested this by running
```
podman --conmon ~/src/conmon/build/conmon exec f36 sh -c 'sleep 1000&'
```

and

```
podman --conmon /usr/bin/conmon exec f36 sh -c 'sleep 1000&'
```

and suspending my laptop.  On resume, `/usr/bin/conmon` was spinning at 100% CPU, but my patched version wasn't.

giuseppe commented on src/ctr_exit.c:
```diff
@@ -83,9 +79,16 @@ gboolean check_child_processes_cb(gpointer user_data)
 }
 
-gboolean on_sigusr1_cb(gpointer user_data)
+gboolean on_signalfd_cb(gint fd, GIOCondition condition, gpointer user_data)
```
`condition` is not used and the CI barks about it (just prepend G_GNUC_UNUSED)

allisonkarlitskaya replied:
Funny.  I had caught this locally as well, and changed it, but forgot to commit before pushing.  Fixed now :)

allisonkarlitskaya commented:
I've added the signed-off-by and fixed the unused parameter warning, as well as updating a comment to be more accurate, and adding an error check if `signalfd()` fails (using the same style as similar checks in the same function).

haircommander commented:
Thanks @allisonkarlitskaya I appreciate you digging in here :smiley: 

```
/usr/bin/ld: src/conmon.o: undefined reference to symbol 'pthread_sigmask@@GLIBC_2.2.5'
/usr/bin/ld: /lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
```
is one failure in ci
maybe we need to include <pthread.h> too?

allisonkarlitskaya commented:
> Thanks @allisonkarlitskaya I appreciate you digging in here smiley
> 
> ```
> /usr/bin/ld: src/conmon.o: undefined reference to symbol 'pthread_sigmask@@GLIBC_2.2.5'
> /usr/bin/ld: /lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
> ```
> 
> is one failure in ci maybe we need to include <pthread.h> too?
> 
> edit: or maybe add `-lpthread` when compiling...

manpage says

```
       #include <signal.h>

       int pthread_sigmask(int how, const sigset_t *set, sigset_t *oldset);

       Compile and link with -pthread.
```

The irony of the fact that I need to link with `-pthread` to avoid creating a thread is not lost on me.

allisonkarlitskaya commented:
So there's also `sigprocmask()` which is basically the same thing.

A bit more reading though, and I ran into this little gem:

```
       A child created via fork(2) inherits a copy of its parent's
       signal mask; the signal mask is  preserved  across execve(2).
```

This isn't fun anymore :sweat_smile:

haircommander commented:
LMK if you'd like one of the conmon devs could take it over :smiley:

allisonkarlitskaya commented:
> LMK if you'd like one of the conmon devs could take it over smiley

I'd be happy to get a bit of input as to if anyone thinks that this is a problem or not.  `conmon` spawns a lot of processes, and having `SIGCHLD` masked in those processes might have all kinds of strange side-effects.

This falls back into the category of "I don't really know enough about what this program is doing"...

giuseppe commented:
> `conmon` spawns a lot of processes, and having `SIGCHLD` masked in those processes might have all kinds of strange side-effects.

I think that is not a problem.  conmon spawns only two processes (the OCI runtime and the cleanup process).  I think it is safe to assume these programs will reset the signals.

allisonkarlitskaya commented:
> > `conmon` spawns a lot of processes, and having `SIGCHLD` masked in those processes might have all kinds of strange side-effects.
> 
> I think that is not a problem. conmon spawns only two processes (the OCI runtime and the cleanup process). I think it is safe to assume these programs will reset the signals.

Good enough for me.  I changed to `sigprocmask()` instead (which is already used from the same file).

haircommander commented:
LGTM, thanks for all of your work @allisonkarlitskaya . @giuseppe for the final approval?

giuseppe reviewed (approved):
LGTM

Jmennius commented:
Can we have this released?
It's been a month since this PR was merged, plus it will take time to got into the distros..
Thanks!

rhatdan commented:
We are preparing releases right now podman 4.2
